hydra-plutus/src/Hydra/Contract/HeadTokens.hs

```diff
@@ -45,29 +46,35 @@ validate initialValidator headValidator seedInput action context =
     Burn -> validateTokensBurning context
 {-# INLINEABLE validate #-}
 
--- FIXME: This doesn't verify that:
---
--- (a) A ST is minted with the right 'HydraHeadV1' name
--- (b) PTs's name have the right shape (i.e. 28 bytes long)
 validateTokensMinting :: ValidatorHash -> ValidatorHash -> TxOutRef -> ScriptContext -> Bool
 validateTokensMinting initialValidator headValidator seedInput context =
   traceIfFalse "minted wrong" $
     participationTokensAreDistributed currency initialValidator txInfo nParties
-      && checkQuantities
-      && assetNamesInPolicy == nParties + 1
+      && checkQuantitiesAndTokenNames
+      && assetNamesInPolicyLength == nParties + 1
       && seedInputIsConsumed
  where
   currency = ownCurrencySymbol context
 
   minted = getValue $ txInfoMint txInfo
 
-  (checkQuantities, assetNamesInPolicy) = case Map.lookup currency minted of
+  (checkQuantitiesAndTokenNames, assetNamesInPolicyLength) = case Map.lookup currency minted of
     Nothing -> (False, 0)
     Just tokenMap ->
       foldr
-        (\q (assertion, n) -> (assertion && q == 1, n + 1))
+        ( \(tn, q) (assertion, n) ->
+            ( assertion
+                && q == 1
+                &&
+                -- Check that we either have 28 bytes long token name for PT
+                -- tokens or in case this is ST token it has the correct name
+                let tnBs = unTokenName tn
+                 in (lengthOfByteString tnBs == 28 || tnBs == Head.hydraHeadV1)
+            , n + 1
+            )
+        )
```

Comment:
```suggestion
        ( \(tokenName, quantity) (assertion, n) ->
            ( assertion
                && quantity == 1
                &&
                -- Check that we either have 28 bytes long token name for PT
                -- tokens or in case this is ST token it has the correct name
                let tokenName' = unTokenName tokenName
                 in (lengthOfByteString tokenName' == 28 || tokenName' == Head.hydraHeadV1)
            , n + 1
            )
        )
```